src/geo.c

```diff
@@ -428,31 +428,56 @@ static int sort_gp_desc(const void *a, const void *b) {
  * Commands
  * ==================================================================== */
 
-/* GEOADD key long lat name [long2 lat2 name2 ... longN latN nameN] */
+/* GEOADD key [CH] [NX|XX] long lat name [long2 lat2 name2 ... longN latN nameN] */
 void geoaddCommand(client *c) {
-    /* Check arguments number for sanity. */
-    if ((c->argc - 2) % 3 != 0) {
+    int options, longidx=2;
+    int i;
+    int xx=0,nx=0;
+    int ns=0; /* not support */
+    char *opt;
+
+    /* We need check the options and locate the longidx. */
+    for (i = 0; i < c->argc-2; i++) {
+        opt = c->argv[longidx]->ptr;
+        if (!strcasecmp(opt, "ch")) {}
+        else if (!strcasecmp(opt, "xx")) xx=1;
+        else if (!strcasecmp(opt, "nx")) nx=1;
+        else if (!strcasecmp(opt, "gt") || !strcasecmp(opt, "lt") ||
+                !strcasecmp(opt, "incr")) {
+            ns=1;
+            break;
+        }
+        else break;
+        longidx++;
+    }
```

Comment:
I find this control structure more approachable. Also, no need to check for options that aren't supported.

```suggestion
    /* Parse options. At the end 'longidx' is set to the argument position
     * of the longitude of the first element. */
    while(longidx < c->argc) {
        char *opt = c->argv[longidx]->ptr;
        if (!strcasecmp(opt,"nx")) nx = 1;
        else if (!strcasecmp(opt,"xx")) xx = 1
        else if (!strcasecmp(opt,"ch")) ch = 1;
        else break;
        longidx++;
    }
```